### PR TITLE
Client Encryption : Add unit tests, integration tests and contract enforcement tests

### LINKED ISSUE #1360: Contracts test for Encryption assembly
Similar to SDK contract JSON, we should have one for Encryption pacakge as well to describe the public contract and prevent inadvertent changes. ref: ContractEnforcement.ContractChanges()

### PR DESCRIPTION
**Pull request by anujtoshniwal**
## Description

Add test coverage.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)

## Closing issues

Closes #1360

### CODE REVIEW

File: Microsoft.Azure.Cosmos.Encryption/src/CosmosDataEncryptionKeyProvider.cs [hunk 1]
```diff
@@ -58,6 +58,11 @@ internal Container Container
                 throw new InvalidOperationException($"{nameof(CosmosDataEncryptionKeyProvider)} has already been initialized.");
             }
 
+            if (database == null)
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/CosmosDataEncryptionKeyProvider.cs:61]: Let's validate containerId as well.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs [hunk 2]
```diff
@@ -143,19 +143,31 @@ public DataEncryptionKeyContainerCore(CosmosDataEncryptionKeyProvider dekProvide
...
+                    cancellationToken);
 
-            return (dekProperties, inMemoryRawDek);
+                return (dekProperties, inMemoryRawDek);
+            }
+            catch (CosmosException ex)
+            {
+                if (ex.Message.Contains("Resource Not Found"))
+                {
+                    throw new ArgumentException($"Data Encryption Key with id: '{id}' not found.");
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:166]: Any specific need to do this? And also should likely not be ArgumentException.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs [hunk 3]
```diff
@@ -65,14 +77,17 @@ internal static class EncryptionProcessor
...
-                JToken propertyValueHolder = itemJObj.Property(propertyName).Value;
+                JProperty propertyValueHolder = itemJObj.Property(propertyName);
 
-                // Even null in the JSON is a JToken with Type Null, this null check is just a sanity check
-                if (propertyValueHolder != null)
+                if (propertyValueHolder == null)
                 {
-                    toEncryptJObj.Add(propertyName, propertyValueHolder.Value<JToken>());
-                    itemJObj.Remove(propertyName);
+                    throw new ArgumentException($"{nameof(encryptionOptions.PathsToEncrypt)} includes a path: '{pathToEncrypt}' not found.");
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs:84]: Add "which was" before "not found" for grammatical correctness.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 2 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:166]: How about new EncryptionKeyNotFoundException ??
Its useful in both read & write paths and also convey the real root of the error.

File: Microsoft.Azure.Cosmos.Encryption/src/CosmosDataEncryptionKeyProvider.cs (hunk 1 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/CosmosDataEncryptionKeyProvider.cs:61]: ArgumentNullException will be thrown from CreateContainerIfNotExistsAsync call if containerId is empty or null.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 2 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs]: Good point, added new EncryptionKeyNotFound exception.

The intention here is to return an exception with a clear message. Went with ArgrurmentException because the issue is due to an input argument provided. Any better alternative?

File: Microsoft.Azure.Cosmos.Encryption/src/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.Azure.Cosmos.Encryption.Tests" + AssemblyKeys.TestPublicKey)]
+[assembly: InternalsVisibleTo("Microsoft.Azure.Cosmos.Encryption.Tests" + AssemblyKeys.ProductPublicKey)]
+
+internal static class AssemblyKeys
+{
+    /// <summary>ProductPublicKey is an official MS supported public key for external releases. TestPublicKey is an unsupported strong key for testing and internal use only</summary>
+    internal const string ProductPublicKey = ", PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9";
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/src/AssemblyInfo.cs:13]: Why is this here? This doesn't exist for the dot net sdk version.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs [hunk 5]
```diff
@@ -143,19 +143,31 @@ public DataEncryptionKeyContainerCore(CosmosDataEncryptionKeyProvider dekProvide
...
+                    this.UnwrapAsync,
+                    diagnosticsContext,
+                    cancellationToken);
 
-            return (dekProperties, inMemoryRawDek);
+                return (dekProperties, inMemoryRawDek);
+            }
+            catch (CosmosException ex)
+            {
+                if (ex.Message.Contains("Resource Not Found"))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:164]: This is to error prone and as a soft contract that can easily be broken. Why is getting the cached dek throwing a cosmos exception instead of just throwing the EncryptionKeyNotFoundException?

Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:164]: I agree this is a soft contract. But, I was thinking we should not throw EncryptionKeyNotFoundException blindly for all underlying exceptions. For other exceptions like ArguementNullException, OperationCancelledException, ObjectDisposedException, OverflowException etc, figured it would be better to relay the CosmosException so that the client can retry / or act accordingly.

File: Microsoft.Azure.Cosmos.Encryption/src/AssemblyInfo.cs (hunk 4 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/AssemblyInfo.cs:13]: Removed.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 5 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:164]: I agree with @j82w , can't we at least use the statuscode/substatuscode or something other than the text?

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionExceptionFactory.cs [hunk 6]
```diff
@@ -35,5 +35,10 @@ internal static Exception InvalidAuthenticationTag()
                 "Invalid authentication tag in cipher text.",
                 "cipherText");
         }
+
+        internal static Exception EncryptionKeyNotFoundException(string encryptionKeyId)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionExceptionFactory.cs:39]: Just curious, if we already know these are ArgumentExceptions, why use Exception as the output type? Will it change in the future?

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs [hunk 7]
```diff
@@ -65,14 +77,17 @@ internal static class EncryptionProcessor
             foreach (string pathToEncrypt in encryptionOptions.PathsToEncrypt)
             {
                 string propertyName = pathToEncrypt.Substring(1);
-                JToken propertyValueHolder = itemJObj.Property(propertyName).Value;
+                JProperty propertyValueHolder = itemJObj.Property(propertyName);
 
-                // Even null in the JSON is a JToken with Type Null, this null check is just a sanity check
-                if (propertyValueHolder != null)
+                if (propertyValueHolder == null)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs:82]: nit: You can probably replace this with a itemJObj.TryGetValue, it will also return the value even if it's null and you don't need the JProperty

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 5 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:164]: Makes sense, checking for StatusCode instead of message.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionExceptionFactory.cs (hunk 6 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionExceptionFactory.cs:39]: good catch, updated the return type to ArguementException.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs (hunk 7 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs:82]: Sure, fixed.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs [hunk 8]
```diff
@@ -143,19 +143,31 @@ public DataEncryptionKeyContainerCore(CosmosDataEncryptionKeyProvider dekProvide
...
+                InMemoryRawDek inMemoryRawDek = await this.DekProvider.DekCache.GetOrAddRawDekAsync(
+                    dekProperties,
+                    this.UnwrapAsync,
+                    diagnosticsContext,
+                    cancellationToken);
 
-            return (dekProperties, inMemoryRawDek);
+                return (dekProperties, inMemoryRawDek);
+            }
+            catch (CosmosException ex)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:162]: Why is this throwing a CosmosException?

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs [hunk 9]
```diff
@@ -64,15 +76,14 @@ internal static class EncryptionProcessor
 
             foreach (string pathToEncrypt in encryptionOptions.PathsToEncrypt)
             {
-                string propertyName = pathToEncrypt.Substring(1);
-                JToken propertyValueHolder = itemJObj.Property(propertyName).Value;
-
-                // Even null in the JSON is a JToken with Type Null, this null check is just a sanity check
-                if (propertyValueHolder != null)
+                string propertyName = pathToEncrypt.Substring(1);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs:79]: What's the reason for the substring? Can you please add a comment on why it exists.

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+
+    [TestCategory("Windows")]
+    [TestClass]
+    public class ContractEnforcement
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs:15]: Can you just file reference the existing file?

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/Microsoft.Azure.Cosmos.Encryption.Tests.csproj [hunk 11]
```diff
@@ -0,0 +1,45 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/Microsoft.Azure.Cosmos.Encryption.Tests.csproj:1]: Do we really need a separate project for unit tests? I understand for larger project, but for the amount of tests it seems like it might be unnecessary.

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/testbaseline.cmd [hunk 12]
```diff
@@ -0,0 +1,113 @@
+@echo off
+
+set COMMAND=
```
Comment (j82w) [inline Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/testbaseline.cmd:3]: Does this file need to be duplicated? Can't it just point the guide to the original file.

Comment (ealsur) [reply Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/testbaseline.cmd:3]: Original file writes on the folder it is running on with the file name already defined.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 8 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:162]: Comes from AsyncCache if resource is not found.

File: Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs (hunk 9 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/EncryptionProcessor.cs:79]: The path starts with '/' (eg "/name"). We have a check for handling the cases where it is not (lines 65-71)

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs (hunk 10 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs:15]: The files aren't exactly the same. The paths (dllName, baseline path, breaking changes path) and the error message differ.

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/Microsoft.Azure.Cosmos.Encryption.Tests.csproj (hunk 11 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/Microsoft.Azure.Cosmos.Encryption.Tests.csproj:1]: Currently the tests are limited, but as we add more functionality there should be more tests.

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs (hunk 10 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs:15]: Can you refactor it into a common helper class that just takes in those few properties to avoid duplicating all this logic?

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 8 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:162]: To be clear the AsyncCache doesn't throw this. This is caused by the DEK no longer existing in the container. I feel like all of these should be wrapped in some form of encryption key failure exception with an inner cosmos exception to make it clear that the operation failed in the get DEK part vs the item operation failing. @kirankumarkolli what are your thoughts?

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:162]: +1 assuming its innerexception ask

File: Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs (hunk 10 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/tests/Microsoft.Azure.Cosmos.Encryption.Tests/ContractEnforcement.cs]: Sure added.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs (hunk 8 above)
Comment (anujtoshniwal) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs]: Makes sense, added the underlying CosmosException as the inner exception for EncryptionKeyNotFound exception we throw.
